ipa-core/src/protocol/context/mod.rs

```diff
@@ -157,37 +160,43 @@ pub trait SpecialAccessToUpgradedContext<F: ExtendableField>: UpgradedContext<F>
 /// Context for protocol executions suitable for semi-honest security model, i.e. secure against
 /// honest-but-curious adversary parties.
 #[derive(Clone)]
-pub struct Base<'a> {
+pub struct Base<'a, B: ShardBinding = NoSharding> {
     inner: Inner<'a>,
     gate: Gate,
     total_records: TotalRecords,
+    sharding: B,
```

Comment:
Does this make more sense as part of `Gateway`, to avoid frequent cloning of static information?

Reply:
I was imagining `Gateway` to be the same for sharded vs non-sharded world because that struct is complicated enough with the stall detection wrapper around it. It would just reference both transports and provide means to open shard-to-shard and mpc channels. 

Depending on the context used, shard-to-shard channels will or will not be be accessible by MPC circuits.

I agree that cloning this stuff often may be less ideal - one way to mitigate that is to move this to `Inner` struct - the downside will be that this will have to be done for all context types separately. I am going to add a comment that we may need to do it later (shouldn't be too hard)

Reply:
I suggested that this also exist on the gateway.

Reply:
I see your point now. Yea, I think I can just return it from shard transport and when creating a gateway for non-sharded execution provide a panicking shard transport to get a runtime error. Maybe parametrizing Gateway can turn this into a compile error, but for some reason I seem to prefer not to do it

Reply:
I started changing Gateway in the next PR, in this change I won't be able to get it, so I'll follow up on it